src/problems/AllPaths/algo/matrix_bfs/intersection.py

```diff
@@ -0,0 +1,189 @@
+            for x, y in product(
+                range(graph_vertices_num), self.regular_automaton.final_states
+            )
+        ]
+
+        for start_state in start_states:
+            enfa.add_start_state(State(start_state))
+
+        for final_state in final_states:
+            enfa.add_final_state(State(final_state))
+
+        for symbol in self.intersection_matrices:
+            matrix = self.intersection_matrices[symbol]
+
+            for row, col in zip(matrix.rows, matrix.cols):
+                enfa.add_transition(State(row), Symbol(symbol), State(col))
+
+        return enfa
+
+    def to_inter_coord(self, graph_vert, reg_vert) -> int:
+        """
+        Converts coordinates of graph vertice and regex vertice
+        to intersection coordinates vertice
+        """
+        return reg_vert * self.graph.get_number_of_vertices() + graph_vert
+
+    def create_diag_matrices(self) -> Dict[str, Matrix]:
+        """
+        Create a block diagonal matrices from graph and regex matrices for each symbol
+        """
+        num_vert_graph = self.graph.get_number_of_vertices()
+        num_vert_regex = self.regular_automaton.num_states
+        diag_num_verts = num_vert_graph + num_vert_regex
+
+        diag_matrices = dict()
+        for symbol in self.regular_automaton.matrices:
+            if symbol in self.graph:
+                diag_matrix = Matrix.sparse(BOOL, diag_num_verts, diag_num_verts)
+                diag_matrix.assign_matrix(
+                    self.regular_automaton.matrices[symbol],
+                    slice(0, num_vert_regex - 1),
+                    slice(0, num_vert_regex - 1),
+                )
+                diag_matrix.assign_matrix(
+                    self.graph[symbol],
+                    slice(num_vert_regex, diag_num_verts - 1),
+                    slice(num_vert_regex, diag_num_verts - 1),
+                )
+
+                diag_matrices[symbol] = diag_matrix
+
+        return diag_matrices
+
+    def intersect_bfs(self) -> EpsilonNFA:
+        """
+        Intersection implementation with synchronous breadth first traversal
+        of a graph and regular grammar represented in automata
+        """
+        num_vert_graph = self.graph.get_number_of_vertices()
+        num_vert_regex = self.regular_automaton.num_states
+
+        num_verts_inter = num_vert_graph * num_vert_regex
+        num_verts_diag = num_vert_graph + num_vert_regex
+
+        graph = self.graph
+        regex = self.regular_automaton.matrices
+
+        regex_start_states = self.regular_automaton.start_states
+
+        diag_matrices = self.create_diag_matrices()
+
+        found = Vector.sparse(BOOL, num_verts_diag)
+        vect = Vector.sparse(BOOL, num_verts_diag)
+
+        # vectors of found nodes
+        # should be done with mask probably, not a queue
+        curr_front = list()
+        visited = Vector.sparse(BOOL, num_verts_inter)
+
+        for regex_start_st in regex_start_states:
```

Comment:
Can we create a multiple-source version instead of multiple running of single-source version?